sycl/source/detail/persistent_device_code_cache.cpp

```diff
@@ -0,0 +1,408 @@
+        isCacheItemSrcEqual(FileName + ".src", Device, Img, SpecConsts,
+                            BuildOptionsString)) {
+      try {
+        return readBinaryDataFromFile(FileName + ".bin");
+      } catch (...) {
+        // If read was unsuccessfull try the next item
+      }
+    }
+    FileName = Path + "/" + std::to_string(++i);
+  }
+  return {};
+}
+
+/* Returns string value which can be used to identify different device
+ */
+std::string PersistentDeviceCodeCache::getDeviceIDString(const device &Device) {
+  return Device.get_platform().get_info<sycl::info::platform::name>() + "/" +
+         Device.get_info<sycl::info::device::name>() + "/" +
+         Device.get_info<sycl::info::device::version>() + "/" +
+         Device.get_info<sycl::info::device::driver_version>();
+}
+
+/* Write built binary to persistent cache
+ * Format: numImages, 1stImageSize, Image[, NthImageSize, NthImage...]
+ * Return on first unsuccessfull file operation
+ */
+void PersistentDeviceCodeCache::writeBinaryDataToFile(
+    const std::string &FileName, const std::vector<std::vector<char>> &Data) {
+  std::ofstream FileStream{FileName, std::ios::binary};
+  if (FileStream.fail())
+    return;
+
+  size_t Size = Data.size();
+  FileStream.write((char *)&Size, sizeof(Size));
+  if (FileStream.fail())
+    return;
+
+  for (size_t i = 0; i < Data.size(); ++i) {
+    Size = Data[i].size();
+    FileStream.write((char *)&Size, sizeof(Size));
+    if (FileStream.fail())
+      return;
+    FileStream.write(Data[i].data(), Size);
+    if (FileStream.fail())
+      return;
+  }
+  FileStream.close();
+}
+
+/* Read built binary to persistent cache
+ * Format: numImages, 1stImageSize, Image[, NthImageSize, NthImage...]
+ */
+std::vector<std::vector<char>>
+PersistentDeviceCodeCache::readBinaryDataFromFile(const std::string &FileName) {
+  std::ifstream FileStream{FileName, std::ios::binary};
+  if (FileStream.fail())
+    return {};
+  size_t ImgNum = 0, ImgSize = 0;
+  FileStream.read((char *)&ImgNum, sizeof(ImgNum));
+  if (FileStream.fail())
+    return {};
+
+  std::vector<std::vector<char>> Res(ImgNum);
+  for (size_t i = 0; i < ImgNum; ++i) {
+    FileStream.read((char *)&ImgSize, sizeof(ImgSize));
+    if (FileStream.fail())
+      return {};
+
+    std::vector<char> ImgData(ImgSize);
+    FileStream.read(ImgData.data(), ImgSize);
+    if (FileStream.fail())
+      return {};
+
+    Res[i] = std::move(ImgData);
+  }
+
+  return Res;
+}
+
+/* Writing cache item key sources to be used for reliable identification
+ * Format: Four pairs of [size, value] for device, build options, specialization
+ * constant values, device code SPIR-V image.
+ */
+void PersistentDeviceCodeCache::writeSourceItem(
+    const std::string &FileName, const device &Device,
+    const RTDeviceBinaryImage &Img, const SerializedObj &SpecConsts,
+    const std::string &BuildOptionsString) {
+  std::ofstream FileStream{FileName, std::ios::binary};
+  if (FileStream.fail())
```

Comment:
I should've probably brought up this earlier, but anyway: this multitude of `if (FileStream.fail()) return;` really hurts readability, frankly. Also, it does not really make sense w/o any error handling.
So please do the check once in the end and somehow let user know that caching failed (e.g. a message to stderr, maybe guarded by env var). This applies to other function with this pattern.

Reply:
done
Enabled by SYCL_CACHE_TRACE function.